File: UVA/AOAPC-BAC/Ch7/Example/UVA-725 Division.py
kase = 0
while True:
    n = int(input().strip())
    if n == 0:
        break
    if kase > 0:
        print()
    kase += 1
    ans = []
    def dfs(i, s, cur):
        if cur * n > 98765: # 剪枝
            return
        if i == 5:
            x = n * cur
            for _ in range(5):
                j = x % 10
                if (1 << j) & s:
                    return
                x //= 10
                s |= 1 << j
            ans.append(cur)
            return
        # 枚舉下一個數字
        for j in range(10):
            if s & (1 << j):
                continue
            dfs(i + 1, s | (1 << j), cur * 10 + j)
    dfs(0, 0, 0)
    if len(ans) == 0:
        print(f"There are no solutions for {n}.")
    else:
        for x in ans:
            print(f"{n * x} / {x:05d} = {n}")

File: UVA/AOAPC-BAC/Ch7/Example/UVA-725 Division.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int n;
vector<int> ans;

void dfs(int s, int cur) {
    if (cur * n > 98765) return; // 剪枝
    if (__builtin_popcount(s) == 5) {
        int x = n * cur;
        for (int j = 0; j < 5; j++) {
            int k = x % 10;
            if (s & (1 << k)) return;
            x /= 10;
            s |= 1 << k;
        }
        ans.push_back(cur);
        return;
    }
    for (int j = 0; j < 10; j++) {
        if (s & (1 << j)) continue;
        dfs(s | (1 << j), cur * 10 + j);
    }
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int kase = 0;
    while (cin >> n && n) {
        if (kase++ > 0) cout << endl;
        ans.clear();
        dfs(0, 0);
        if (ans.empty()) {
            cout << "There are no solutions for " << n << "." << endl;
        }
        else {
            for (int x : ans) {
                cout << x * n << " / " << setw(5) << setfill('0') << x << " = " << n << endl;
            }
        }
    }
    return 0;
}